### ISSUE TITLE
Coredumping on FreeBSD 10.2 amd64 with clang-ed 2.3.0

### ISSUE THREAD

**Issue by stargrave** [PurpleI2P/i2pd#333]
I have got FreeBSD 10.2 amd64 with clang-compiled 2.3.0. It was running for several days and some traffic was passing by. But I stopped the daemon and tried to run it again and it coredumps:

    $ ./i2pd --v6=1 --bandwidth=P --floodfill=1 --port=9000 --daemon=0
    i2pd v2.3.0 starting

    00:27:42/debug - data directory: /home/stargrave/.i2pd
    00:27:42/debug - Daemon: CMD parameters:
    00:27:42/debug - 0:  ./i2pd
    00:27:42/debug - 1:  --v6=1
    00:27:42/debug - 2:  --bandwidth=P
    00:27:42/debug - 3:  --floodfill=1
    00:27:42/debug - 4:  --port=9000
    00:27:42/debug - 5:  --daemon=0
    00:27:42/info - Log: min msg level set to info
    00:27:42/info - Daemon: staring HTTP Server
    00:27:42/info - Daemon: starting NetDB
    00:27:42/info - NetDb: 3871 routers loaded (1217 floodfils)
    00:27:42/info - Daemon: starting Transports
    00:27:42/info - NTCP: Start listening TCP port 9000
    Abort trap (core dumped)

    $ truss ./i2pd --v6=1 --bandwidth=P --floodfill=1 --port=9000 --daemon=0
    [...]
    NTCP: Start listening TCP port 9000
    write(2,"NTCP: Start listening TCP port 9"...,36) = 36 (0x24)
    _umtx_op(0x805c17880,UMTX_OP_MUTEX_WAIT,0x0,0x0,0x0) = 0 (0x0)
    _umtx_op(0x805c17830,UMTX_OP_MUTEX_WAIT,0x0,0x0,0x0) = 0 (0x0)
    _umtx_op(0x805c17830,UMTX_OP_MUTEX_WAKE2,0x0,0x0,0x0) = 0 (0x0)
    getpid()                                         = 37875 (0x93f3)
    ioctl(13,FIONBIO,0xffffcfac)                     = 0 (0x0)
    kevent(10,{0xd,EVFILT_READ,EV_ADD|EV_CLEAR,0,0x0,0x805cc70e0},1,0x0,0,0x0) = 0 (0x0)
    socket(PF_INET6,SOCK_STREAM,6)                   = 14 (0xe)
    setsockopt(0xe,0xffff,0x800,0x7fffffffd010,0x4,0x806b5a440) = 0 (0x0)
    getpid()                                         = 37875 (0x93f3)
    setsockopt(0xe,0x29,0x1b,0x7fffffffd388,0x4,0x4) = 0 (0x0)
    bind(14,{ AF_INET6 [a0ff:bd00:800:0:6fe6:ffff:100:0]:36096 },28) ERR#48 'Address already in use'
    sigprocmask(SIG_SETMASK,SIGHUP|SIGINT|SIGQUIT|SIGILL|SIGTRAP|SIGEMT|SIGFPE|SIGKILL|SIGBUS|SIGSEGV|SIGSYS|SIGPIPE|SIGALRM|SIGTERM|SIGURG|SIGSTOP|SIGTSTP|SIGCONT|SIGCHLD|SIGTTIN|SIGTTOU|SIGIO|SIGXCPU|SIGXFSZ|SIGVTALRM|SIGPROF|SIGWINCH|SIGINFO|SIGUSR1|SIGUSR2,0x0) = 0 (0x0)
    thr_self(0x7fffffffd000,0x7fffffffd000,0x0,0x0,0x7fffffffcbd8,0x803a00000) = 0 (0x0)
    thr_kill(0x1874b,0x6,0x0,0x0,0x7fffffffcbd8,0x803a00000) = 0 (0x0)
    SIGNAL 6 (SIGABRT)
    sigprocmask(SIG_SETMASK,SIGHUP|SIGINT|SIGQUIT|SIGILL|SIGTRAP|SIGABRT|SIGEMT|SIGFPE|SIGBUS|SIGSEGV|SIGSYS|SIGPIPE|SIGALRM|SIGTERM|SIGURG|SIGTSTP|SIGCONT|SIGCHLD|SIGTTIN|SIGTTOU|SIGIO|SIGXCPU|SIGXFSZ|SIGVTALRM|SIGPROF|SIGWINCH|SIGINFO|SIGUSR1|SIGUSR2,0x0)
    = 0 (0x0)
    sigreturn(0x7fffffffc600,0x7fffffffc600,0x301,0x0,0xfffffffffffffbc0,0x803a00000) = 0 (0x0)
    sigprocmask(SIG_SETMASK,SIGHUP|SIGINT|SIGQUIT|SIGILL|SIGTRAP|SIGABRT|SIGEMT|SIGFPE|SIGKILL|SIGBUS|SIGSEGV|SIGSYS|SIGPIPE|SIGALRM|SIGTERM|SIGURG|SIGSTOP|SIGTSTP|SIGCONT|SIGCHLD|SIGTTIN|SIGTTOU|SIGIO|SIGXCPU|SIGXFSZ|SIGVTALRM|SIGPROF|SIGWINCH|SIGINFO|SIGUSR1|SIGUSR2,SIGHUP|SIGINT|SIGQUIT|SIGILL|SIGTRAP|SIGEMT|SIGFPE|SIGBUS|SIGSEGV|SIGSYS|SIGPIPE|SIGALRM|SIGTERM|SIGURG|SIGTSTP|SIGCONT|SIGCHLD|SIGTTIN|SIGTTOU|SIGIO|SIGXCPU|SIGXFSZ|SIGVTALRM|SIGPROF|SIGWINCH|SIGINFO|SIGUSR1|SIGUSR2) = 0 (0x0)
    sigaction(SIGABRT,{ SIG_DFL 0x0 ss_t },{ 0x8029ac080 SA_RESTART|SA_SIGINFO ss_t }) = 0 (0x0)
    sigprocmask(SIG_SETMASK,SIGHUP|SIGINT|SIGQUIT|SIGILL|SIGTRAP|SIGEMT|SIGFPE|SIGBUS|SIGSEGV|SIGSYS|SIGPIPE|SIGALRM|SIGTERM|SIGURG|SIGTSTP|SIGCONT|SIGCHLD|SIGTTIN|SIGTTOU|SIGIO|SIGXCPU|SIGXFSZ|SIGVTALRM|SIGPROF|SIGWINCH|SIGINFO|SIGUSR1|SIGUSR2,0x0) = 0 (0x0)
    sigprocmask(SIG_SETMASK,SIGHUP|SIGINT|SIGQUIT|SIGILL|SIGTRAP|SIGEMT|SIGFPE|SIGKILL|SIGBUS|SIGSEGV|SIGSYS|SIGPIPE|SIGALRM|SIGTERM|SIGURG|SIGSTOP|SIGTSTP|SIGCONT|SIGCHLD|SIGTTIN|SIGTTOU|SIGIO|SIGXCPU|SIGXFSZ|SIGVTALRM|SIGPROF|SIGWINCH|SIGINFO|SIGUSR1|SIGUSR2,0x0) = 0 (0x0)
    thr_self(0x7fffffffd000,0x7fffffffd000,0x0,0x0,0x7fffffffcbd8,0x803a00000) = 0 (0x0)
    thr_kill(0x1874b,0x6,0x0,0x0,0x7fffffffcbd8,0x803a00000) = 0 (0x0)
    SIGNAL 6 (SIGABRT)
    _umtx_op(0x800abe008,UMTX_OP_WAIT_UINT_PRIVATE,0x0,0x0,0x0) ERR#4 'Interrupted system call'
    process exit, rval = 0

But if I try to run on different port (for example 9001), then it starts and works. I do not well understand that dump, but seems that "Address already in use" is a bad thing. And there is strangeness: my daemon is off for several minutes, "ps" does not show any i2p daemon running, but there is some busy TCP connection on its port:

    # sockstat | grep 9000
    ?        ?          ?     ?  tcp6   2001:470:1f0b:10ad:fe:4aff:fec8:9c00:9000 2001:0:53aa:64c:387a:2667:ac77:e2f2:49626

Possibly exactly it means that some resources are busy.

**Comment by orignal** [comment]
I see the same problem with v6 at linux. Probably I don't close a socket properly.
If you know the reason let me know.

**Comment by stargrave** [comment]
After sometime, my still alive connection timeouted and disappeared -- I was able to start the daemon again.

Unfortunately I am not sure, as I know C and low-level network programming rather bad.

I saw similar behaviour either when working with sockets in Python, or when invoking "socat": if I did not specify "socket reuse" option while dealing with it, then when program is finished,
socket stays somewhat alive and, when program starts again, it fails to bind on it with "already in use" error. There is similar issue, I assume.
